rest/oidc_api_test.go

```diff
@@ -1,139 +1,714 @@
+			redirectURI := locationURL.Query().Get(requestParamRedirectURI)
+			redirectURL, err := url.Parse(redirectURI)
+			require.NoError(t, err, "error parsing redirect_uri URL")
+			assert.Equal(t, tc.wantProvider, redirectURL.Query().Get(auth.OIDCAuthProvider))
+
+			// Check _oidc behavior
+			authURL = strings.ReplaceAll(tc.authURL, "${path}", "_oidc")
+			resp = rt.SendAdminRequest(http.MethodGet, authURL, "")
+			require.Equal(t, http.StatusFound, resp.Code)
+			location = resp.Header().Get("Location")
+
+			require.NotEmpty(t, location, "error extracting location from header")
+			locationURL, err = url.Parse(location)
+			require.NoError(t, err, "error parsing location URL")
+			redirectURI = locationURL.Query().Get(requestParamRedirectURI)
+			redirectURL, err = url.Parse(redirectURI)
+			require.NoError(t, err, "error parsing redirect_uri URL")
+			assert.Equal(t, tc.wantProvider, redirectURL.Query().Get(auth.OIDCAuthProvider))
+		})
+	}
+}
+
+// Returns a new OIDCProvider.
+func mockProvider(name string) *auth.OIDCProvider {
+	return &auth.OIDCProvider{
+		Name:          name,
+		ClientID:      base.StringPointer("baz"),
+		UserPrefix:    name,
+		ValidationKey: base.StringPointer("qux"),
+	}
+}
+
+// Returns an auto registration enabled provider.
+func mockProviderWithRegister(name string) *auth.OIDCProvider {
+	return &auth.OIDCProvider{
+		Name:          name,
+		ClientID:      base.StringPointer("baz"),
+		UserPrefix:    name,
+		ValidationKey: base.StringPointer("qux"),
+		Register:      true,
+	}
+}
+
+// Returns a new OIDCProvider with Register and IncludeAccessToken flags enabled.
+func mockProviderWithRegisterWithAccessToken(name string) *auth.OIDCProvider {
+	return &auth.OIDCProvider{
+		Name:               name,
+		ClientID:           base.StringPointer("baz"),
+		UserPrefix:         name,
+		ValidationKey:      base.StringPointer("qux"),
+		Register:           true,
+		IncludeAccessToken: true,
+	}
+}
+
+// Checks End to end OpenID Connect Authorization Code flow.
+func TestOpenIDConnectAuth(t *testing.T) {
+	type test struct {
+		name              string
+		providers         auth.OIDCProviderMap
+		defaultProvider   string
+		authURL           string
+		forceAuthError    forceError
+		forceRefreshError forceError
+	}
+	tests := []test{
+		{
+			// Successful new user authentication when auto registration is enabled and a single provider
+			// is configured through providers configuration. Explicitly specified to include access token in
+			// token response.
+			name: "successful user registration against single provider",
+			providers: auth.OIDCProviderMap{
+				"foo": mockProviderWithRegisterWithAccessToken("foo"),
+			},
+			defaultProvider: "foo",
+			authURL:         "/db/_oidc?provider=foo&offline=true",
+		}, {
+			// Unsuccessful new user authentication when auto registration is enabled and a single provider
+			// is configured through providers configuration. Explicitly specified to include access token in
```

Comment:
Is this correct?  I don't see the access token being enabled here, or auto-register enabled.

Reply:
This is not correct. Fixed comments in latest revision.